src/components/Firestore/Requests/RequestDetails/CodeViewer/index.tsx

```diff
@@ -0,0 +1,119 @@
+/**
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import 'codemirror/keymap/sublime';
+import 'codemirror/theme/xq-light.css';
+
+import './index.scss';
+
+import { ThemeProvider } from '@rmwc/theme';
+import CodeMirror from '@uiw/react-codemirror';
+import React, { useCallback, useEffect, useState } from 'react';
+
+import { errorTheme, successTheme } from '../../../../../themes';
+import { OutcomeInfo, RulesOutcome } from '../../rules_evaluation_result_model';
+import { ICON_SELECTOR } from '../../utils';
+import CodeViewerAdminRequest from './AdminRequest';
+
+interface Props {
+  firestoreRules?: string;
+  linesOutcome?: OutcomeInfo[];
+  isAdminRequest: boolean;
+}
+
+const RulesCodeViewer: React.FC<Props> = ({
+  firestoreRules,
+  linesOutcome,
+  isAdminRequest,
+}) => {
+  const [
+    codeMirrorEditor,
+    setCodeMirrorEditor,
+  ] = useState<CodeMirror.Editor | null>(null);
+
+  // Creates HTMLElement of a Marker containing the outcome icon
+  const getMarkerElement = useCallback((outcome: RulesOutcome): HTMLElement => {
```

Comment:
If I read this correctly, this function doesn't access any variables in this component -- you may as well just move it to module scope and remove the `useCallback` wrapper (since it will be effectively constant).